### PR TITLE
Fixed #30761 -- Prevented floatformat filter from returning a negative zero.

### PR DESCRIPTION
**Pull request by countermeasure**
Old behaviour:
>>> floatformat(-0.001, 2)
'-0.00'
>>> floatformat(-0.4, 0)
'-0'

New behaviour:
>>> floatformat(-0.001, 2)
'0.00'
>>> floatformat(-0.4, 0)
'0'

### CODE REVIEW

File: django/template/defaultfilters.py [hunk 1]
```diff
@@ -154,10 +154,11 @@ def floatformat(text, arg=-1):
     # and `exponent` from Decimal.as_tuple() directly.
     sign, digits, exponent = d.quantize(exp, ROUND_HALF_UP, Context(prec=prec)).as_tuple()
     digits = [str(digit) for digit in reversed(digits)]
+    number_is_not_zero = (set(digits) != {'0'})
     while len(digits) <= abs(exponent):
         digits.append('0')
     digits.insert(-exponent, '.')
-    if sign:
+    if sign and number_is_not_zero:
```
Comment (codingjoe) [inline django/template/defaultfilters.py]:
```suggestion
    if sign and d:
```

File: django/template/defaultfilters.py [hunk 2]
```diff
@@ -154,10 +154,11 @@ def floatformat(text, arg=-1):
     # and `exponent` from Decimal.as_tuple() directly.
     sign, digits, exponent = d.quantize(exp, ROUND_HALF_UP, Context(prec=prec)).as_tuple()
     digits = [str(digit) for digit in reversed(digits)]
+    number_is_not_zero = (set(digits) != {'0'})
```
Comment (codingjoe) [inline django/template/defaultfilters.py]:
```suggestion
```

**Comment by codingjoe** [comment]
Oh and before I forget, you should add a changelog entry to the 3.1 release 👍

**Comment by countermeasure** [comment]
Thanks for the speedy review, and for the suggestion to use `Decimal.__bool__` rather than the set, @codingjoe.

I couldn't use `d` in the `if` statement, as it's the pre-rounding value, so I made `rounded_d` and used that.

What do you think?

File: docs/releases/3.1.txt [hunk 3]
```diff
@@ -279,6 +279,9 @@ Miscellaneous
 * ``django.utils.decorators.classproperty()`` decorator is moved to
   ``django.utils.functional.classproperty()``.
 
+* The :tfilter:`floatformat` filter no longer prefixes non-zero numbers which
```
Comment (codingjoe) [inline docs/releases/3.1.txt]: Maybe rearrange the sentence a little, something like:
```
* :tfilter:`floatformat` no longer adds minus prefix to zero rounded from negative values.
```

**Comment by felixxm** [comment]
@countermeasure Thanks for this patch :+1: Welcome aboard :boat: 

I moved unrelated tests to a separate commit, pushed minor edits and fixed release notes.

**Comment by countermeasure** [comment]
Thanks @codingjoe and @felixxm.

Those changes look good to me :+1:

File: docs/releases/3.1.txt [hunk 4]
```diff
@@ -279,6 +279,9 @@ Miscellaneous
 * ``django.utils.decorators.classproperty()`` decorator is moved to
   ``django.utils.functional.classproperty()``.
 
+* :tfilter:`floatformat` template filter no longer returns a negative zero
+  ``-0`` for negative numbers which are rounded to zero.
+
```
Comment (carltongibson) [inline docs/releases/3.1.txt]: Maybe focusing on the _positive_ (😀) ...

```
* :tfilter:`floatformat` template filter now outputs (postive) ``0`` for 
  negative numbers which round exactly to zero.
```

(But...)

**Comment by felixxm** [comment]
buildbot

**Pull request merged**